src/mutable-prim-array-types.hs

```diff
@@ -31,74 +35,106 @@ unsafeCopy ::
     -> Int -- ^ offset into source array
     -> Int -- ^ number of elements to copy
     -> m ()
-unsafeCopy (Array dst#) (I# doff#) (Array src#) (I# soff#) (I# n#)
-  = primitive_ (copyMutableByteArray#
-      src#
-      (soff# *# (sizeOf# (undefined :: a)))
-      dst#
-      (doff# *# (sizeOf# (undefined :: a)))
-      (n# *# (sizeOf# (undefined :: a)))
-    )
+unsafeCopy (Array dst#) (I# doff#) (Array src#) (I# soff#) (I# n#) =
+    let toBytes cnt# = cnt# *# (sizeOf# (undefined :: a))
+     in primitive_ $
+            copyMutableByteArray#
+                src# (toBytes soff#) dst# (toBytes doff#) (toBytes n#)
 
+-------------------------------------------------------------------------------
+-- Length
+-------------------------------------------------------------------------------
+
+-- XXX rename to byteCount?
+{-# INLINE byteLength #-}
+byteLength :: Array s a -> Int
+byteLength (Array arr#) = I# (sizeofMutableByteArray# arr#)
+
+-- XXX Rename length to elemCount so that there is no confusion bout what it
+-- means.
+--
+-- XXX Since size of 'a' is statically known, we can replace `quot` with shift
+-- when it is power of 2. Though it may not matter unless length is used too
+-- often.
+--
 {-# INLINE length #-}
-length ::
-       forall s a. Prim a
-    => Array s a
-    -> Int
-length (Array arr#) =
-    I# (quotInt# (sizeofMutableByteArray# arr#) (sizeOf# (undefined :: a)))
+length :: forall s a. Prim a => Array s a -> Int
+length arr = byteLength arr `quot` (sizeOf (undefined :: a))
 
-{-# INLINE spliceTwo #-}
-spliceTwo ::
+-------------------------------------------------------------------------------
+-- Random Access
+-------------------------------------------------------------------------------
+
+-- XXX Rename to unsafeReadIndex
+-- Check correctness
+{-# INLINE unsafeIndexM #-}
+unsafeIndexM ::
        forall m a. (PrimMonad m, Prim a)
     => Array (PrimState m) a
-    -> Array (PrimState m) a
-    -> m (Array (PrimState m) a)
-spliceTwo a1 a2 = do
-    a3 <- resizeArray a1 (l1 + l2)
-    unsafeCopy a2 0 a3 l1 l2
-    return a3
-  where
-    l1 = length a1
-    l2 = length a2
+    -> Int
+    -> m a
+unsafeIndexM (Array arr#) (I# i#) =
+    -- XXX XXX why not use readByteArray#
```

Comment:
Will change, this isn't a bug though, is it?

Reply:
The state token is not used during the read.

Reply:
The idea is that I can only read from a mutable byte given its current state after changes etc. Else, I think the results wont be as expected due to optimizations. Eitherway, we have `readByteArray#` now.